### ISSUE TITLE
Parse Server 2.2.25 and file upload -> {"error":"unauthorized"}

### ISSUE THREAD

**Issue by gateway** [ParsePlatform/parse-php-sdk#270]
It seems that in Parse 2.2.24 and above this addition was added to parse server "Better support for checking application and client keys, thanks to Steven Shipton"

My file uploads stopped working, in 2.2.23 they work fine but anything above that I get unauthorized.

Here is the line of code that workes in 2.2.23

```
   // save file to Parse
    $parseFile = ParseFile::createFromData(file_get_contents($uri), $name, $type);
    $parseFile->save($useMasterKey = true);
```

When tracing thought he issue in the php sdk it seems that masterkey is never set in _getRequestHeaders function, its empty.

The first debug point is from _getRequestHeaders in the ParseClient file, and the one below that is the one in the ParseFile section.. its missing masterkey 

![image](https://cloud.githubusercontent.com/assets/519563/20979848/1993a550-bc63-11e6-9677-554e6eb78ada.png)

If I manually set $useMasterKey = TRUE in _getRequestHeaders it works.

Any ideas? I think the server is working like it should and being more secure but the php sdk should be updated??

**Comment by gateway** [comment]
So I did a quick fix if anyone is running into this, I'm not sure if this is the *RIGHT* way but it works now.

https://github.com/ParsePlatform/parse-php-sdk/blob/master/src/Parse/ParseFile.php#L188-L239

which is 

```
    /**
     * Uploads the file contents to Parse, if not saved.
     *
     * @return bool
     */
    public function save()
    {
        if (!$this->url) {
            $response = $this->upload();
            $this->url = $response['url'];
            $this->name = $response['name'];
        }
        return true;
    }
    private function upload()
    {
        $fileParts = explode('.', $this->getName());
        $extension = array_pop($fileParts);
        $mimeType = $this->mimeType ?: $this->getMimeTypeForExtension($extension);
        $headers = ParseClient::_getRequestHeaders(null, false);
        $url = ParseClient::getAPIUrl().'files/'.$this->getName();
        $rest = curl_init();
        curl_setopt($rest, CURLOPT_URL, $url);
        curl_setopt($rest, CURLOPT_RETURNTRANSFER, 1);
        curl_setopt($rest, CURLOPT_BINARYTRANSFER, 1);
        $headers[] = 'Content-Type: '.$mimeType;
        curl_setopt($rest, CURLOPT_POST, 1);
        curl_setopt($rest, CURLOPT_POSTFIELDS, $this->getData());
        curl_setopt($rest, CURLOPT_HTTPHEADER, $headers);
        $response = curl_exec($rest);
        $contentType = curl_getinfo($rest, CURLINFO_CONTENT_TYPE);
        if (curl_errno($rest)) {
            throw new ParseException(curl_error($rest), curl_errno($rest));
        }
        curl_close($rest);
        if (strpos($contentType, 'text/html') !== false) {
            throw new ParseException('Bad Request', -1);
        }
        $decoded = json_decode($response, true);
        if (isset($decoded['error'])) {
            throw new ParseException(
                $decoded['error'],
                isset($decoded['code']) ? $decoded['code'] : 0
            );
        }
        return $decoded;
    }
```

to

```
    /**
     * Uploads the file contents to Parse, if not saved.
     *
     * @return bool
     */
    public function save($useMasterKey = false)
    {
        if (!$this->url) {
            $response = $this->upload($useMasterKey);
            $this->url = $response['url'];
            $this->name = $response['name'];
        }

        return true;
    }

    private function upload($useMasterKey = false)
    {
        $fileParts = explode('.', $this->getName());
        $extension = array_pop($fileParts);
        $mimeType = $this->mimeType ?: $this->getMimeTypeForExtension($extension);
        $headers = ParseClient::_getRequestHeaders(null, $useMasterKey);
        $url = ParseClient::getAPIUrl().'files/'.$this->getName();
        $rest = curl_init();
        curl_setopt($rest, CURLOPT_URL, $url);
        curl_setopt($rest, CURLOPT_RETURNTRANSFER, 1);
        curl_setopt($rest, CURLOPT_BINARYTRANSFER, 1);
        $headers[] = 'Content-Type: '.$mimeType;
        curl_setopt($rest, CURLOPT_POST, 1);
        curl_setopt($rest, CURLOPT_POSTFIELDS, $this->getData());
        curl_setopt($rest, CURLOPT_HTTPHEADER, $headers);
        $response = curl_exec($rest);
        $contentType = curl_getinfo($rest, CURLINFO_CONTENT_TYPE);
        if (curl_errno($rest)) {
            throw new ParseException(curl_error($rest), curl_errno($rest));
        }
        curl_close($rest);
        if (strpos($contentType, 'text/html') !== false) {
            throw new ParseException('Bad Request', -1);
        }
        $decoded = json_decode($response, true);
        if (isset($decoded['error'])) {
            throw new ParseException(
                $decoded['error'],
                isset($decoded['code']) ? $decoded['code'] : 0
            );
        }

        return $decoded;
    }
```

For those who cant spot the changes here is a diff.

![image](https://cloud.githubusercontent.com/assets/519563/20981772/cc7e1a68-bc6a-11e6-8f76-71a9a6592bb3.png)

@gfosco @GrahamCampbell @yaman-jain @juliangut any thoughts?

**Comment by montymxb** [comment]
This is interesting...I took a look at all the existing calls to ```ParseClient::_getRequestHeaders``` to compare, turns out there are only 3 in the parse php sdk. Two of these are in ParseFile, respectively ```ParseFile::upload``` and ```ParseFile::delete```. 

The final call is in ```ParseClient::_request``` which takes ```$useMasterKey``` as a parameter from it's caller, almost all requests to the backend go through this, that is except for the 2 in ParseFile.

 I think you may have stumbled across something that should probably be changed. It seems odd that ParseFile should have it's own set of curl calls for talking to the server, and with hardcoded values for ```$useMasterKey``` as well. I'm not sure as to what changed between 2.2.23 and 2.2.24 in Parse server to cause your issue to come up, but it seems to have made this underlying implementation a rather glaring issue.

For now your fix is probably fine, but I would recommend you change ```ParseFile::delete``` as well on your end to take a bool for master key use as well (it's currently hardcoded as ```true```). I'll see if I can't put together a pull request swapping those independent curl requests for some standard calls to ```ParseClient::_request```.

**Comment by gateway** [comment]
Well im glad I could point this out, ill do some manual fixes for now until maybe someone updates the sdk.

**Comment by montymxb** [comment]
Just merged in #274 , this should take care of your issue. I'll leave this open until you can confirm the patch works.
